tactics/class_tactics.ml

```diff
@@ -1227,18 +1230,18 @@ let resolve_all_evars depth unique env p oevd fail =
           (* evars_to_goals p evd gives none when there's no evar having p *)
           match evars_to_goals p evd with
           | Some (goals, nongoals) ->
-            let solver = find_solver env evd comp in
-            let evm, goalsls = Search.preprocess_goals evd goals in
-            let finished, evm' = solver env evm ~goals:goalsls ~best_effort:true ~depth ~unique in
-            let evm' = Search.post_process_goals ~goals ~nongoals ~evm ~evm' ~finished in
-            if has_undefined p oevd evm' then
+            let { solver } = find_solver env evd comp in
+            let evd, old_metas, sorted_goals = Search.preprocess_goals evd goals in
+            let finished, evd = solver env evd ~goals:sorted_goals ~best_effort:true ~depth ~unique in
```

Comment:
Last nitpick, I promise: for even more robustness you should perform the projection on this line, as
```ocaml
            let solver = find_solver env evd comp in
            let evd, old_metas, sorted_goals = Search.preprocess_goals evd goals in
            let finished, evd = solver.solver env evd ~goals:sorted_goals ~best_effort:true ~depth ~unique in
```

Reply:
No, problem :-)
I've commited right now what you suggest

Reply:
I'm extremely wary of the mess introduced by optional and named arguments. Having a statically known arity at apply time is always good to drive the unification errors into a readable output. (Also there might be considerations about functions being fully applied.)